### DISCUSSION TITLE
Nifti Images with SIRF

### DISCUSSION THREAD

**Discussion by Kiarashdnsh** [SyneRBI/SIRF#1281] [Q&A] [answered]
Hello!

I'm trying to reconstruct PET images in nifti format using the SIRF VM. Is this even possible? since in the examples provided, only the interfile hv and hs formats have been used. I also changed the nifti format to h33, i33, and v formats using the XMedcon application.
However, the code I use
```python
image_data = pet.ImageData('path/to/file/m000-31013_AV451_S_E.v')
``` 
renders this error:
```python
ERROR: Error opening file /media/sf_DATABASE/m000-m000-31013_AV451_S_E.v
``` 
and that
```python
the reconstruction engine output may provide more information
``` 
I get the same error message when using h33/i33 as well.

If you could help in this matter, I would much appreciate it. Thanks!

**Comment by paskino** [comment]
Hi @Kiarashdnsh sorry for late reply. You should pass a path to the interfile header file to the `pet.ImageData` creator. From your example, it seems you are passing the binary data.

However, you want to load a NIFTI file, this should doable as:

```python
from sirf import Reg
from sirf import STIR as pet

data = Reg.ImageData('/path/to/data.nii')

petdata = pet.ImageData(data)
```

**Reply by Kiarashdnsh** [reply]
Hello @paskino 
Thanks for your reply. I still have problems with this issue, so even late replies are most welcome!

I did what you said. I'm getting this error:

```python
---------------------------------------------------------------------------
error                                     Traceback (most recent call last)
Cell In[7], line 5
      2 from sirf import STIR as pet
      4 data = Reg.ImageData('/media/sf_PET-MR/Test Dataset/Images/31000_AV45_S_E.nii')
----> 5 petdata = pet.ImageData(data)

File ~/devel/install/python/sirf/STIR.py:429, in ImageData.__init__(self, arg)
    427         raise AssertionError()
    428     self.handle = pystir.cSTIR_imageFromImageData(arg.handle)
--> 429     check_status(self.handle)
    430 elif arg is not None:
    431     raise error(
    432         'wrong argument ' + repr(arg) + ' for ImageData constructor')

File ~/devel/install/python/sirf/Utilities.py:451, in check_status(handle, stack)
    446 line = pyiutil.executionErrorLine(handle)
    447 errorMsg = \
    448     repr(msg) + ' exception caught at line ' + \
    449     repr(line) + ' of ' + file + '; ' + \
    450     'the reconstruction engine output may provide more information'
--> 451 raise error(errorMsg)

error: ??? "'TODO - create STIRImageData from general SIRFImageData.' exception caught at line 1264 of /home/sirfuser/devel/buildVM/sources/SIRF/src/xSTIR/cSTIR/cstir.cpp; the reconstruction engine output may provide more information"
``` 

I've also attached the `cstir.cpp` file.
[cstir.zip](https://github.com/user-attachments/files/16872008/cstir.zip)

Thanks

**Comment by evgueni-ovtchinnikov** [comment]
so far as I remember you were using SIRF 3.5 - have you tried version 3.8?

**Reply by Kiarashdnsh** [reply]
Yes, I have tried updating, however, I could not update to 3.8, getting this message in the process:
```python
[ 31%] Completed 'GTest'
[ 31%] Built target GTest
[ 32%] Performing download step (git clone) for 'JSON'
Cloning into 'JSON'...
error: 32752 bytes of body are still expected
fetch-pack: unexpected disconnect while reading sideband packet
fatal: early EOF
fatal: fetch-pack: invalid index-pack output
...
...
...
Cloning into 'range-v3'...
error: RPC failed; curl 92 HTTP/2 stream 0 was not closed cleanly: CANCEL (err 8)
error: 4106 bytes of body are still expected
fetch-pack: unexpected disconnect while reading sideband packet
fatal: early EOF
fatal: fetch-pack: invalid index-pack output

-- Had to git clone more than once: 3 times.

CMake Error at /home/sirfuser/devel/buildVM/builds/JSON/tmp/JSON-gitclone.cmake:39 (message):
Failed to clone repository: 'https://github.com/nlohmann/json'

gmake[2]: *** [CMakeFiles/JSON.dir/build.make:99: builds/JSON/stamp/JSON-download] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:398: CMakeFiles/JSON.dir/all] Error 2
gmake[1]: *** Waiting for unfinished jobs....
error: RPC failed; curl 92 HTTP/2 stream 0 was not closed cleanly: CANCEL (err 8)
error: 120 bytes of body are still expected
fetch-pack: unexpected disconnect while reading sideband packet
fatal: early EOF
fatal: fetch-pack: invalid index-pack output

-- Had to git clone more than once: 3 times.

CMake Error at /home/sirfuser/devel/buildVM/builds/range-v3/tmp/range-v3-gitclone.cmake:39 (message):
Failed to clone repository: 'https://github.com/ericniebler/range-v3'

gmake[2]: *** [CMakeFiles/range-v3.dir/build.make:99: builds/range-v3/stamp/range-v3-download] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:372: CMakeFiles/range-v3.dir/all] Error 2
gmake: *** [Makefile:101: all] Error 2
```

**Reply by KrisThielemans** [reply]
By the way, this is likely due to temporary networking problems, either your side or GitHub. You could try again, also reduce the number of parallel builds if necessary (although I don't think that'll be it).

**Reply by Kiarashdnsh** [reply]
Thanks, @KrisThielemans, I'll try again with a VPN.

**Comment by KrisThielemans** [comment, marked as answer]
There's a few things here (all suggestoins only work if you compiled STIR with ITK):
- "I'm trying to reconstruct PET images in nifti format using the SIRF VM. " The rest of the question is actually not related to that, as you're only asking about reading. The answer to your original question is "yes". Likely something like
  ```python
   im = recon.get_output()
   regim = reg.ImageData(im)
   regim.write('bla.nii')
   ```
   The (better) alternative is to use `sirf.STIR` to write directly as Nifti, without going through our conversion. This would use the `im.write_par` functionality. Hopefully, there's an example...
- The error you're getting about the conversion is still in 3.8 AFAIK. But you wouldn't need it:
  ```python
  im = pet.ImageData('/media/sf_PET-MR/Test Dataset/Images/31000_AV45_S_E.nii')
  ```

- Sadly, it looks like we haven't uploaded a 3.8 VM yet, but you could try to [update yours](https://github.com/SyneRBI/SIRF-SuperBuild/blob/master/VirtualBox/documentation/Basic-usage.md#updating-your-vm). I'd highly recommend making a copy first though (or at least a snapshot).

**Reply by danieldeidda** [reply]
The generate_NEMA_ROI in contribs has examples of how to use im.write_par 

https://github.com/SyneRBI/SIRF-Contribs/blob/master/src/Python/sirf/contrib/NEMA/generate_nema_rois.py#L165-L167

**Reply by Kiarashdnsh** [reply]
Thanks @KrisThielemans 

```python
im = pet.ImageData('/media/sf_PET-MR/Test Dataset/Images/31000_AV45_S_E.nii')
``` 
saved the day.
